lib/msf/core/post/linux/compile.rb

```diff
@@ -1,63 +1,101 @@
 # -*- coding: binary -*-
+
 module Msf
-class Post
-module Linux
-module Compile
-  include ::Msf::Post::Common
-  include ::Msf::Post::File
-  include ::Msf::Post::Unix
-
-  def initialize(info = {})
-    super
-    register_options( [
-      OptEnum.new('COMPILE', [true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']]),
-    ], self.class)
-  end
-
-  def live_compile?
-    return false unless %w{ Auto True }.include?(datastore['COMPILE'])
-
-    if has_gcc?
-      vprint_good 'gcc is installed'
-      return true
-    end
-
-    unless datastore['COMPILE'] == 'Auto'
-      fail_with Module::Failure::BadConfig, 'gcc is not installed. Set COMPILE False to upload a pre-compiled executable.'
-    end
-  end
-
-  def upload_and_compile(path, data, gcc_args='')
-    write_file "#{path}.c", strip_comments(data)
-
-    gcc_cmd = "gcc -o '#{path}' '#{path}.c'"
-    if session.type == 'shell'
-      gcc_cmd = "PATH=\"$PATH:/usr/bin/\" #{gcc_cmd}"
-    end
-
-    unless gcc_args.to_s.blank?
-      gcc_cmd << " #{gcc_args}"
-    end
-
-    output = cmd_exec gcc_cmd
-    rm_f "#{path}.c"
-
-    unless output.blank?
-      print_error output
-      message = "#{path}.c failed to compile."
-      # don't mention the COMPILE option if it was deregistered
-      message << ' Set COMPILE to False to upload a pre-compiled executable.' if options.include?('COMPILE')
-      fail_with Module::Failure::BadConfig, message
-    end
-
-    chmod path
-  end
-
-  def strip_comments(c_code)
-    c_code.gsub(%r{/\*.*?\*/}m, '').gsub(%r{^\s*//.*$}, '')
-  end
-
-end # Compile
-end # Linux
-end # Post
+  class Post
+    module Linux
+      module Compile
+        include ::Msf::Post::Common
+        include ::Msf::Post::File
+        include ::Msf::Post::Unix
+
+        def initialize(info = {})
+          super
+          register_options([
+            OptEnum.new('COMPILE', [true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']]),
+            OptString.new('CC', [true, 'The compiler to use', 'gcc'], conditions: ['COMPILE', 'in', ['Auto', 'True']]),
+            OptString.new('MAKE', [true, 'The build system to use', 'make'], conditions: ['COMPILE', 'in', ['Auto', 'True']])
+          ], self.class)
+        end
+
+        def live_compile?
+          return false unless %w[Auto True].include?(datastore['COMPILE'])
+
+          if command_exists?(datastore['CC']) || command_exists?(datastore['MAKE'])
+            vprint_good "#{datastore['CC']} is installed"
+            vprint_good "#{datastore['MAKE']} is installed"
+            return true
+          end
```

Comment:
The logic of this method no longer makes sense.

The `if` conditional uses `||`. If either the compiler or make are installed, the method will suggest that both are installed.